scarb/src/internal/serdex.rs

```diff
@@ -1,9 +1,73 @@
-use anyhow::Result;
+use anyhow::{anyhow, Result};
 use camino::{Utf8Path, Utf8PathBuf};
 use serde::{Deserialize, Serialize};
 
 use crate::internal::fsx;
 
+const MERGE_STRATEGY_KEY: &str = "merge-strategy";
+
+#[derive(Debug, Default)]
+pub enum MergeStrategy {
+    #[default]
+    Override,
+    Merge,
+}
+
+impl TryFrom<&str> for MergeStrategy {
+    type Error = anyhow::Error;
+    fn try_from(value: &str) -> std::result::Result<Self, Self::Error> {
+        match value {
+            "override" => Ok(Self::Override),
+            "merge" => Ok(Self::Merge),
+            _ => Err(anyhow!(
+                "invalid merge strategy: {}, must be one of `merge`, `override`",
+                value
+            )),
+        }
+    }
+}
+
+pub fn toml_merge_apply_strategy<'de, T, S>(target: &T, source: &S) -> Result<T>
```

Comment:
great idea! 🔥 
I wonder though if this function could be refactored a bit, I'm generally not a fan of doing multiple nested loops. Maybe in order to improve readability and maintainability you could do something like (please treat it as a pseudocode as I have not tested this):

```
    let mut target_value = Value::try_from(target)?;
    let source_value = Value::try_from(source)?;
    
    if let (Some(target_table), Some(source_table)) = (target_value.as_table_mut(), source_value.as_table()) {
        for (key, value) in source_table {
            match target_table.get_mut(key) {
                Some(target_subvalue) if target_subvalue.is_table() && value.is_table() => {
                    let target_subtable = target_subvalue.as_table_mut().unwrap();
                    let value_subtable = value.as_table().unwrap();

                    let strategy = value_subtable.get(MERGE_STRATEGY_KEY)
                        .and_then(|v| v.as_str())
                        .map_or(Ok(MergeStrategy::default()), MergeStrategy::try_from)?;

                    match strategy {
                        MergeStrategy::Override => {
                            *target_subvalue = value_subtable.clone();
                        }
                        MergeStrategy::Merge => {
                            *target_subvalue = toml_merge_apply_strategy(target_subtable, value_subtable)?;
                        }
                    }
                }
                _ => {
                    target_table.insert(key.clone(), value.clone());
                }
            }
        }
    }

    Ok(toml::Value::try_into(target_value)?)
```
So we could get rid of a few ifs there, and flatten the code a bit. Let me know what you think, though you may treat this comment as a nitpick, as I'm not sure if this fits in your (scarb) code-style yet 😅

Reply:
Surprisingly, making stuff simpler instead of more complicated does fit the Scarb code-style 🤣

Reply:
![Alt Text](https://media1.tenor.com/images/432ebaaed3a1edb68a41e3423c287979/tenor.gif?itemid=7250508)